### PR TITLE
Feature/export cluster hierarchy

### PR DESCRIPTION
**Pull request by CanArisan**
⚠️ Depends on ls1intum/Athene#26 to be deployed as well! (DONE)

---

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] ~Client: I added multiple integration tests (Jest) related to the features~
- [x] ~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~
- [x] ~Client: I documented the TypeScript code using JSDoc style.~
- [x] ~Client: I added multiple screenshots/screencasts of my UI changes~
- [x] ~Client: I translated all the newly inserted strings into German and English~

### Motivation and Context

To offer automatic feedback to text blocks which were:
- Not assigned to any cluster in the initial clustering
- Was assigned to a cluster, but it does not have an existing feedback yet

The UML class diagram and the new database tables can be seen in this issue:
[https://github.com/ls1intum/Artemis/issues/1782](url)

### Description

- Extended the response from TextSimilarityClusteringService to include the clusterTree and distanceMatrix of blocks
- Moved all text exercise related domain classes to a new "text" package
- Added corresponding entities and repositories
     - TextTreeNode
     - TextTreeNodeRepository
     - TextPairwiseDistance
     - TextPairwiseDistanceRepository
- Modified TextSimilarityClusteringServiceTest to mock the new response of Athene
- Added TextClusteringServiceTest to test the new operations on the database

### Testing
Before testing make sure that the latest version of **Athene is running** with all of its components!
There is **no new feature** added in this PR, only new database entities and repositories are added, and some existing ones are slightly modified. The new test TextClusteringServiceTest.java covers all queries in the new repositories. 

**Steps for Testing Locally:**
1. Start Athene using "docker-compose up -d" 
2. Run Artemis Server (automaticText enabled) and Client
3. Create a text exercise with automatic feedback suggestions enabled and due date a couple of minutes later
4. Submit a text submission with >= 10 sentences (a couple of paragraphs copied from wikipedia etc.)
5. Wait for the due date of the exercise to pass so that clustering can be performed.
6. Check the database tables: text_block, text_cluster, text_pairwise_distance, text_tree_node should have corresponding entries

### Test Coverage
service.TextClusteringService - 92% (Previously 44%)

domain.text.TextTreeNode - 83% (New Entity)
domain.text.TextPairwiseDistance  - 83% (New Entity)
domain.text.TextExercise - 93% (Previously 92%)
domain.text.TextBlock - 90% (Previously 89%)
domain.text.TextCluster - 73% (Previously 70%)

### CODE REVIEW

**Comment by krusche** [comment]
I would like to discuss the relevant database changes in https://github.com/ls1intum/Artemis/issues/1782 before we review and test this PR. Due to the current extremely high workload, I can only offer you a meeting on July 28 or afterwards.

Please make sure that all information in https://github.com/ls1intum/Artemis/issues/1782 is structured in an easy to understand way in the main post (at the top) using headlines and bullet points.

File: src/main/java/de/tum/in/www1/artemis/domain/ClusteringResult.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
+package de.tum.in.www1.artemis.domain;
+
+import java.io.Serializable;
+import java.util.List;
+
+import javax.persistence.*;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/ClusteringResult.java:6]: Codacy found an issue: [Avoid unused imports such as 'javax.persistence'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/test/java/de/tum/in/www1/artemis/service/connectors/TextSimilarityClusteringServiceTest.java [hunk 2]
```diff
@@ -8,8 +8,10 @@
 import java.io.IOException;
 import java.net.HttpURLConnection;
 import java.net.URL;
+import java.util.ArrayList;
```
Comment (jpbernius) [inline src/test/java/de/tum/in/www1/artemis/service/connectors/TextSimilarityClusteringServiceTest.java:11]: Codacy found an issue: [Avoid unused imports such as 'java.util.ArrayList'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/repository/TreeNodeRepository.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+import de.tum.in.www1.artemis.domain.TextExercise;
+import de.tum.in.www1.artemis.domain.TreeNode;
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+import static org.springframework.data.jpa.repository.EntityGraph.EntityGraphType.LOAD;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/repository/TreeNodeRepository.java:11]: Codacy found an issue: [Avoid unused imports such as 'org.springframework.data.jpa.repository.EntityGraph.EntityGraphType.LOAD'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/service/AutomaticTextFeedbackService.java [hunk 4]
```diff
@@ -84,4 +111,106 @@ public void suggestFeedback(@NotNull Result result) {
...
+            List<TreeNode> siblings = treeNodeRepository.findAllByParentAndExercise(parentId, exercise);
+            siblings.sort(Comparator.comparingDouble(TreeNode::getLambda_val));
+            // In the following loop, we trace all ancestors within the threshold of the siblings for feedback
+            while(!siblings.isEmpty()) {
+                TreeNode x = siblings.remove(0);
+                // If already below lambda threshold remove x directly
+                if(addLambdaValues(x.getLambda_val(), currentLambdaVal) < LAMBDA_THRESHOLD) {
+                    continue;
+                }
+                if(x.isBlockNode()) {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/AutomaticTextFeedbackService.java:139]: Codacy found an issue: [Avoid empty if statements](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/TreeNode.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    public long getParent() { return parent; }
+
+    public void setParent(long parent) { this.parent = parent; }
+
+    public long getChild() { return child; }
+
+    public void setChild(long child) { this.child = child; }
+
+    public double getLambda_val() {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/TreeNode.java:55]: Codacy found an issue: [The instance method name 'getLambda_val' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/TextExercise.java [hunk 6]
```diff
@@ -24,6 +24,21 @@
...
 
+    /*
+    @OneToOne(fetch = FetchType.LAZY)
+    @JsonIgnore
+    private ClusteringResult clusteringResult;
+     */
+
+    @OneToMany(mappedBy = "exercise")
+    @JsonIgnore
+    private List<PairwiseDistance> pairwiseDistances;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/TextExercise.java:35]: Codacy found an issue: [Avoid unused private fields such as 'pairwiseDistances'.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/ClusteringResult.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
+package de.tum.in.www1.artemis.domain;
+
+import java.io.Serializable;
+import java.util.List;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/ClusteringResult.java:4]: Codacy found an issue: [Avoid unused imports such as 'java.util.List'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/PairwiseDistance.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
+package de.tum.in.www1.artemis.domain;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import org.hibernate.annotations.NaturalId;
+
+import java.io.Serializable;
+
+import javax.persistence.*;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/PairwiseDistance.java:8]: Codacy found an issue: [Avoid unused imports such as 'javax.persistence'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/repository/TextBlockRepository.java [hunk 9]
```diff
@@ -7,6 +7,7 @@
 import java.util.Set;
 
 import org.springframework.data.jpa.repository.*;
+import org.springframework.data.repository.query.Param;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/repository/TextBlockRepository.java:10]: Codacy found an issue: [Avoid unused imports such as 'org.springframework.data.repository.query.Param'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/repository/PairwiseDistanceRepository.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
+package de.tum.in.www1.artemis.repository;
+
+import de.tum.in.www1.artemis.domain.TextExercise;
+import de.tum.in.www1.artemis.domain.PairwiseDistance;
+import org.springframework.data.jpa.repository.EntityGraph;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/repository/PairwiseDistanceRepository.java:5]: Codacy found an issue: [Avoid unused imports such as 'org.springframework.data.jpa.repository.EntityGraph'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/ClusteringResult.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+import javax.persistence.*;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/ClusteringResult.java:8]: Codacy found an issue: [Avoid unused imports such as 'com.fasterxml.jackson.annotation.JsonIgnore'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/TreeNode.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        return lambda_val == -1 ? Double.POSITIVE_INFINITY : lambda_val;
+    }
+
+    public void setLambda_val(double lambda_val) { this.lambda_val = lambda_val; }
+
+    public long getChild_size() { return child_size; }
+
+    public void setChild_size(long child_size) { this.child_size = child_size; }
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/TreeNode.java:63]: Codacy found an issue: [The instance method name 'setChild_size' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/ClusteringResult.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
+package de.tum.in.www1.artemis.domain;
+
+import java.io.Serializable;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/ClusteringResult.java:3]: Codacy found an issue: [Avoid unused imports such as 'java.io.Serializable'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/TreeNode.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
+package de.tum.in.www1.artemis.domain;
+
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import org.hibernate.annotations.NaturalId;
+
+import javax.persistence.*;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/TreeNode.java:7]: Codacy found an issue: [Avoid unused imports such as 'javax.persistence'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/service/AutomaticTextFeedbackService.java [hunk 15]
```diff
@@ -30,11 +32,25 @@
...
 
+    // TODO: Update this value later
+    private static final double LAMBDA_THRESHOLD = 3;
+
     private final TextBlockRepository textBlockRepository;
 
-    public AutomaticTextFeedbackService(FeedbackService feedbackService, TextBlockRepository textBlockRepository) {
+    private final TreeNodeRepository treeNodeRepository;
+
+    private final PairwiseDistanceRepository pairwiseDistanceRepository;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/AutomaticTextFeedbackService.java:42]: Codacy found an issue: [Avoid unused private fields such as 'pairwiseDistanceRepository'.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/PairwiseDistance.java [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+
+    public long getId() {
+        return id;
+    }
+
+    public void setId(long id) {
+        this.id = id;
+    }
+
+    public long getBlock_i() { return this.block_i; }
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/PairwiseDistance.java:46]: Codacy found an issue: [The instance method name 'getBlock_i' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/repository/TextBlockRepository.java [hunk 17]
```diff
@@ -24,5 +25,17 @@
...
+    @Query("SELECT distinct b FROM TextBlock b " +
+        "LEFT JOIN FETCH Submission s on b.submission = s " +
+        "LEFT JOIN Participation p ON s.participation = p " +
+        "LEFT JOIN TextExercise e ON p.exercise = e " +
+        "WHERE e.id = :#{#exerciseId} AND b.treeId IS NOT NULL")
+    List<TextBlock> findAllByTreeIdExistsAndExerciseId(@Param("id") Long exerciseId);
+     */
+
+    // TODO: Test this query
+    List<TextBlock> findAllBySubmission_Participation_Exercise_IdAndTreeIdNotNull(Long exerciseId);
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/repository/TextBlockRepository.java:38]: Codacy found an issue: [The instance method name 'findAllBySubmission_Participation_Exercise_IdAndTreeIdNotNull' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/PairwiseDistance.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
+package de.tum.in.www1.artemis.domain;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import org.hibernate.annotations.NaturalId;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/PairwiseDistance.java:4]: Codacy found an issue: [Avoid unused imports such as 'org.hibernate.annotations.NaturalId'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/test/java/de/tum/in/www1/artemis/service/connectors/TextSimilarityClusteringServiceTest.java [hunk 19]
```diff
@@ -42,18 +64,25 @@ public void clusterTextBlocks() throws NetworkingError {
...
-        assertThat(distanceMatrix[0][1], is(equalTo(distanceMatrix[1][0])));
-        assertThat(distanceMatrix[0][1], is(both(greaterThan(0.5)).and(lessThan(0.7))));
+        List<List<Double>> matrix = response.distanceMatrix;
+        assertThat(matrix, hasSize(blocks.size()));
+
+        List<TreeNode> clusterTree = response.clusterTree;
+        List<TreeNode> blocksInTree = clusterTree.stream().filter(treeNode -> treeNode.isBlockNode()).collect(Collectors.toList());
+        // Assert that number of blockNodes in the tree equals number of blocks
+        assertThat(blocksInTree, hasSize(blocks.size()));
+        List<Long> groupByChild = clusterTree.stream().map(treeNode -> treeNode.getChild()).distinct().collect(Collectors.toList());
```
Comment (jpbernius) [inline src/test/java/de/tum/in/www1/artemis/service/connectors/TextSimilarityClusteringServiceTest.java:80]: Codacy found an issue: [Unnecessary use of fully qualified name 'Collectors.toList' due to existing static import 'java.util.stream.Collectors.toList'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/TreeNode.java [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+
+    public long getChild() { return child; }
+
+    public void setChild(long child) { this.child = child; }
+
+    public double getLambda_val() {
+        return lambda_val == -1 ? Double.POSITIVE_INFINITY : lambda_val;
+    }
+
+    public void setLambda_val(double lambda_val) { this.lambda_val = lambda_val; }
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/TreeNode.java:59]: Codacy found an issue: [The instance method name 'setLambda_val' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/TreeNode.java [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+
+    public long getChild_size() { return child_size; }
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/TreeNode.java:61]: Codacy found an issue: [The instance method name 'getChild_size' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/PairwiseDistance.java [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+        this.id = id;
+    }
+
+    public long getBlock_i() { return this.block_i; }
+
+    public long getBlock_j() { return block_j; }
+
+    public void setBlock_i(long block_i) { this.block_i = block_i; }
+
+    public void setBlock_j(long block_j) { this.block_j = block_j; }
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/PairwiseDistance.java:52]: Codacy found an issue: [The instance method name 'setBlock_j' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/domain/PairwiseDistance.java [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+
+    public void setId(long id) {
+        this.id = id;
+    }
+
+    public long getBlock_i() { return this.block_i; }
+
+    public long getBlock_j() { return block_j; }
+
+    public void setBlock_i(long block_i) { this.block_i = block_i; }
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/PairwiseDistance.java:50]: Codacy found an issue: [The instance method name 'setBlock_i' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/main/java/de/tum/in/www1/artemis/service/AutomaticTextFeedbackService.java (hunk 15 above)
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/AutomaticTextFeedbackService.java:42]: Codacy found an issue: [Perhaps 'pairwiseDistanceRepository' could be replaced by a local variable.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

File: src/test/java/de/tum/in/www1/artemis/service/connectors/TextSimilarityClusteringServiceTest.java [hunk 24]
```diff
@@ -42,18 +64,25 @@ public void clusterTextBlocks() throws NetworkingError {
...
+        // Assert that child is a unique property of a TreeNode
+        assertThat(groupByChild, hasSize(clusterTree.size()));
+        List<Long> groupByParent = clusterTree.stream().map(treeNode -> treeNode.getParent()).distinct().collect(Collectors.toList());
```
Comment (jpbernius) [inline src/test/java/de/tum/in/www1/artemis/service/connectors/TextSimilarityClusteringServiceTest.java:83]: Codacy found an issue: [Unnecessary use of fully qualified name 'Collectors.toList' due to existing static import 'java.util.stream.Collectors.toList'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968857)

**Comment by jpbernius** [comment]
@actions Tests are failing due to dependency to ls1intum/Athene#26.

**Review by krusche — CHANGES_REQUESTED** [review]
Please make sure to mock the communication to Athene in the server integration tests.
Please post the test coverage for the changed files in the PR description.
Please make sure to cover all cases for the deletion of related entities (course, exam, exercise, exercise group, participation, submission, result) so that we do not have any problems in those cases

**Comment by krusche** [comment]
> Please make sure to mock the communication to Athene in the server integration tests.
> Please post the test coverage for the changed files in the PR description.
> Please make sure to cover all cases for the deletion of related entities (course, exam, exercise, exercise group, participation, submission, result) so that we do not have any problems in those cases

@CanArisan Please let us know, if you addressed those change requests. In particular, whether you have been able to mock all communication with Athene in the tests and whether the deletion is tested thoroughly

**Comment by CanArisan** [comment]
> > Please make sure to mock the communication to Athene in the server integration tests.
> > Please post the test coverage for the changed files in the PR description.
> > Please make sure to cover all cases for the deletion of related entities (course, exam, exercise, exercise group, participation, submission, result) so that we do not have any problems in those cases
> 
> @CanArisan Please let us know, if you addressed those change requests. In particular, whether you have been able to mock all communication with Athene in the tests and whether the deletion is tested thoroughly

I test the deletion for the relevant cases for the new entities that I created, which are course, exam, exercise group and exercise.
I also mocked the communication to Athene in the new test that I created (TextClusteringServiceTest). There is however a test already in develop (TextSimilarityClusteringServiceTest) which communicates with Athene. I did not mock it there because it only tests the response from Athene and does not make sense when mocked. If we want absolutely no communication to Athene, I would remove that test entirely.

**Comment by krusche** [comment]
> > > Please make sure to mock the communication to Athene in the server integration tests.
> > > Please post the test coverage for the changed files in the PR description.
> > > Please make sure to cover all cases for the deletion of related entities (course, exam, exercise, exercise group, participation, submission, result) so that we do not have any problems in those cases
> > 
> > 
> > @CanArisan Please let us know, if you addressed those change requests. In particular, whether you have been able to mock all communication with Athene in the tests and whether the deletion is tested thoroughly
> 
> I test the deletion for the relevant cases for the new entities that I created, which are course, exam, exercise group and exercise.
> I also mocked the communication to Athene in the new test that I created (TextClusteringServiceTest). There is however a test already in develop (TextSimilarityClusteringServiceTest) which communicates with Athene. I did not mock it there because it only tests the response from Athene and does not make sense when mocked. If we want absolutely no communication to Athene, I would remove that test entirely.

1. Can you point to the place where you test the deletion?
2. Why would it not make sense? You would still invoke the code that is responsible for sending a request and handling the response. We can still make sure that there are no exceptions in the Artemis side. Please also mock this one.

**Comment by CanArisan** [comment]
> > > > Please make sure to mock the communication to Athene in the server integration tests.
> > > > Please post the test coverage for the changed files in the PR description.
> > > > Please make sure to cover all cases for the deletion of related entities (course, exam, exercise, exercise group, participation, submission, result) so that we do not have any problems in those cases
> > > 
> > > 
> > > @CanArisan Please let us know, if you addressed those change requests. In particular, whether you have been able to mock all communication with Athene in the tests and whether the deletion is tested thoroughly
> > 
> > 
> > I test the deletion for the relevant cases for the new entities that I created, which are course, exam, exercise group and exercise.
> > I also mocked the communication to Athene in the new test that I created (TextClusteringServiceTest). There is however a test already in develop (TextSimilarityClusteringServiceTest) which communicates with Athene. I did not mock it there because it only tests the response from Athene and does not make sense when mocked. If we want absolutely no communication to Athene, I would remove that test entirely.
> 
> 1. Can you point to the place where you test the deletion?
> 2. Why would it not make sense? You would still invoke the code that is responsible for sending a request and handling the response. We can still make sure that there are no exceptions in the Artemis side. Please also mock this one.

1. The following test cases in src/test/java/de/tum/in/www1/artemis/service/TextClusteringServiceTest.java are testing the deletion:
    - line 272: testExerciseRemoval
    - line 287: testCourseRemoval
    - line 307: testExamRemoval
    - line 327: testExerciseGroupRemoval
2. I also mocked the communication to Athene in the existing test class.

**Comment by krusche** [comment]
@CanArisan what is the status of this PR? It is open since 3 months now and it is getting outdated.
Should we close it, or can you finish the PR in the next days? This would be really important, because our policies normally prohibit PRs that are open more than 2 weeks.

**Review by bigultekin — APPROVED** [review]
I tested locally and observed that all the tables mentioned in step 6 are filled. Also, automatic feedback suggestion and feedback conflicts are still working too. Good job! 👍

File: src/main/java/de/tum/in/www1/artemis/domain/text/TextPairwiseDistance.java [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+import javax.persistence.*;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import de.tum.in.www1.artemis.domain.DomainObject;
+
+/**
+ *  Pairwise distance between two TextBlocks
+ */
+@Entity
+@Table(name = "text_pairwise_distance")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/text/TextPairwiseDistance.java:12]: the following annotations are missing

@Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
@JsonInclude(JsonInclude.Include.NON_EMPTY)

File: src/main/java/de/tum/in/www1/artemis/domain/text/TextPairwiseDistance.java [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+
+    public TextExercise getExercise() {
+        return exercise;
+    }
+
+    public void setExercise(TextExercise exercise) {
+        this.exercise = exercise;
+    }
+
+    public TextPairwiseDistance exercise(TextExercise exercise) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/text/TextPairwiseDistance.java:60]: please delete this method in case it is unused

**Pull request closed without merging**